### ISSUE TITLE
bravado-core does not support multiple parameters properly

### ISSUE THREAD

**Issue by nickdirienzo** [Yelp/bravado-core#258]
We're experiencing an issue where `cast_request_param` is failing with a `TypeError` in the following scenario. For the record, we use pyramid_swagger and by virtue of that, bravado-core.

If you have a spec like:
```
"/objects":
    get:
      operationId: list_objects
      tags:
        - Events
      parameters:
        - in: query
          name: parent_id
          required: true
          description: The ID of the Parent you would like to list all Objects for
          type: integer
          format: int64
```

Then make a request with:
```
GET /resource?parent_id=123&parent_id=456
```

bravado-core will raise a `TypeError` and therefore a 500 because we're trying to cast `['123', '456']` to via `int()`.

I may be misunderstanding how [parameter descriptions](https://swagger.io/docs/specification/2-0/describing-parameters/) work in Swagger 2.0, but from my perspective, this seems like a problem in the library.

A simple fix to not return 500s, would be to also catch `TypeError` and then allow jsonschema to handle the validation (as suggested by the comment [here](https://github.com/Yelp/bravado-core/blob/54d099e3fa1df0fe027908e96dd3718b11b689bf/bravado_core/param.py#L257)), which would then would return a 400 instead. This seems more proper. I'll put up a PR to address the issue, but happy to discuss further in case I'm mistaken.

Looking forward to hearing from you @sjaensch!

**Comment by nickdirienzo** [comment]
Hmm, thinking about this a bit more, I'm not sure if this is an issue with bravado-core or pyramid_swagger given that the docstring of `cast_request_param` is:
```
Try to cast a request param (e.g. query arg, POST data) from a string to
    its specified type in the schema. This allows validating non-string params.

:param param_type: name of the type to be casted to
    :type  param_type: string
    :param param_name: param name
    :type  param_name: string
    :param param_value: param value
    :type  param_value: string
```

So we shouldn't ever be sending a list into this function in the first place. I think I need to dig a bit more here.

**Comment by nickdirienzo** [comment]
Ah okay. This seems like that's the root cause now. As long as the `cast_param` function doesn't raise an exception we will eventually hit the collection format line [here](https://github.com/Yelp/bravado-core/blob/54d099e3fa1df0fe027908e96dd3718b11b689bf/bravado_core/param.py#L194), which handles collections properly. Unfortunately, we're raising a `TypeError` and bombing out before we should.

Collection evaluation looks to be in the right place as long as the casting function doesn't fail exceptionally. Unfortunately, it seems that the docstring may also be wrong as it can support lists as well as strings.

**Comment by macisamuele** [comment]
@nickdirienzo I think that you have a small issues on your swagger specs.
``parent_id`` is defined as an ``integer`` but you are actually sending a list of integers, so this led me think that ``parent_id`` should be an array of integers.
```yaml
"/objects":
    get:
      operationId: list_objects
      tags:
        - Events
      parameters:
        - in: query
          name: parent_id
          required: true
          description: The ID of the Parent you would like to list all Objects for
          collectionFormat: multi  #
          type: array              #
          items:                   #
              type: integer        #
              format: int64        #
```

Closing it for now. Feel free to reopen the issue if your problem is not solved.

**Comment by nickdirienzo** [comment]
@macisamuele thanks for the reply! So we'd like for the query parameter to not operate as a collection, i.e. not support `parent_id=123&parent_id=456` right now. This seems like a valid use of Swagger and to support a collection in the request with multiple query parameters, we'd have to set the `collectionFormat` to `multi` as you suggested.

In this world where we don't want to permit multiple values, we'd like to return a 400 to the consumers instead of a 500, which seems like the proper response because the consumer is not using the interface properly. Is this a correct understanding?

If so, presently, bravado-core causes a 500 instead of a 400 because of the `TypeError` that is raised by calling `int()`. Or am I still misunderstanding here?

**Comment by nickdirienzo** [comment]
What I'm proposing is to allow jsonschema to handle the error instead of bailing out, e.g. like https://github.com/optimizely/bravado-core/commit/07bf13cde0e17395fc81ec4ab95bddf4621343d6. With this approach, instead of a 500, we would get a 400 back with a message of `"Invalid value for field 'parent_id': ['123', '456'] is not of type 'integer'"`. Does this seem like a reasonable thing to want / do?

**Comment by macisamuele** [comment]
I'm reopening this as it didn't solved the issue.
I would ask you more information about the exception that pyramid is reporting (stacktrace will be appreciated).

**Comment by nickdirienzo** [comment]
Definitely can provide you with that! Sorry I didn't provide that originally and instead only provided my understanding of the stack trace:

```
2018-03-02 10:24:46,744 ERROR [frontdoor.models.core:712][MainThread] int() argument must be a string, a bytes-like object or a number, not 'list'
Traceback (most recent call last):
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/pyramid-1.6.1-py3.6.egg/pyramid/tweens.py", line 20, in excview_tween
    response = handler(request)
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/pyramid_swagger-2.3.2-py3.6.egg/pyramid_swagger/tween.py", line 179, in validator_tween
    op_or_validators_map,
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/pyramid_swagger-2.3.2-py3.6.egg/pyramid_swagger/tween.py", line 466, in _validate
    return f(*args, **kwargs)
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/pyramid_swagger-2.3.2-py3.6.egg/pyramid_swagger/tween.py", line 574, in swaggerize_request
    request_data = unmarshal_request(request, op)
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/bravado_core/request.py", line 67, in unmarshal_request
    param_value = unmarshal_param(param, request)
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/bravado_core/param.py", line 171, in unmarshal_param
    raw_value = cast_param(request.query.get(param.name, default_value))
  File "/Users/ndirienzo/.virtualenvs/frontdoor/lib/python3.6/site-packages/bravado_core/param.py", line 253, in cast_request_param
    try:
TypeError: int() argument must be a string, a bytes-like object or a number, not 'list'
```

**Comment by sjaensch** [comment]
Fix released in 4.13.1. Thanks!

**Comment by nickdirienzo** [comment]
You're welcome! Thanks again for publishing the new version so quickly.
